lasagne/tests/test_nonlinearities.py

```diff
@@ -1,17 +1,50 @@
+import pytest
 import numpy as np
+import theano.tensor as T
 
 
-def test_rectify():
-    from lasagne.nonlinearities import rectify
-    assert [rectify(x) for x in (-1, 0, 1, 2)] == [0, 0, 1, 2]
+class TestNonlinearities(object):
+    def linear(self, x):
+        return x
 
+    def rectify(self, x):
+        return x * (x > 0)
 
-def test_leaky_rectify():
-    from lasagne.nonlinearities import LeakyRectify
-    lr = LeakyRectify(0.1)
-    assert np.allclose(lr(np.array([-1, 0, 1, 2])), [-.1, 0, 1, 2])
+    def leaky_rectify(self, x):
+        return x * (x > 0) + 0.01 * x * (x < 0)
 
+    def leaky_rectify_0(self, x):
+        return self.rectify(x)
 
-def test_linear():
-    from lasagne.nonlinearities import linear
-    assert [linear(x) for x in (-1, 0, 1, 2)] == [-1, 0, 1, 2]
+    def sigmoid(self, x):
+        return 1 / (1 + np.exp(-x))
+
+    def tanh(self, x):
+        return np.tanh(x)
+
+    def softmax(self, x):
+        return (np.exp(x).T / np.exp(x).sum(-1)).T
+
+    @pytest.mark.parametrize('nonlinearity',
+                             ['linear', 'rectify',
+                              'leaky_rectify', 'sigmoid',
+                              'tanh', 'softmax',
+                              'leaky_rectify_0'])
+    def test_nonlinearity(self, nonlinearity):
+        import lasagne.nonlinearities
+
+        if nonlinearity == 'leaky_rectify_0':
+            from lasagne.nonlinearities import LeakyRectify
+            theano_nonlinearity = LeakyRectify(leakiness=0)
+        else:
+            theano_nonlinearity = getattr(lasagne.nonlinearities,
+                                          nonlinearity)
+        np_nonlinearity = getattr(self, nonlinearity)
+
+        X = T.matrix()
+        X0 = lasagne.utils.floatX(np.random.uniform(-3, 3, (10, 10)))
+
+        theano_result = theano_nonlinearity(X).eval({X: X0})
+        np_result = np_nonlinearity(X0)
+
+        assert np.allclose(theano_result, np_result)
```

Comment:
:+1: for generalizing the code a bit. Also the previous version relied on the functions working for nonsymbolic input, which is not actually what we needed to test.